azure_functions_worker/shared_memory_data_transfer/file_accessor_unix.py

```diff
@@ -0,0 +1,159 @@
+
+    def _create_mem_map_dir(self) -> bool:
+        """
+        Create a directory to create memory maps.
+        Returns True if either a valid directory already exists or one was
+        created successfully, False otherwise.
+        """
+        # Iterate over all the possible directories where the memory map could
+        # be created and try to create in one of them.
+        for mem_map_temp_dir in consts.UNIX_TEMP_DIRS:
+            dir_path = os.path.join(mem_map_temp_dir,
+                                    consts.UNIX_TEMP_DIR_SUFFIX)
+            if os.path.isdir(dir_path):
+                # One of the directories already exists, no need
+                return True
+            try:
+                os.makedirs(dir_path)
+                return True
+            except Exception:
+                # We try to create a directory in each of the applicable
+                # directory paths until we successfully create one or one that
+                # already exists is found.
+                # Even if this fails, we keep trying others.
+                pass
+        # Could not create a directory in any of the applicable directory paths.
+        # We will not be able to create any memory maps so we fail.
+        logger.error('Cannot create directory for memory maps')
+        return False
+
+    def _create_mem_map_file(self, mem_map_name: str, mem_map_size: int) \
+            -> Optional[BufferedRandom]:
+        """
+        Get the file descriptor for a new memory map.
+        Returns the file descriptor.
+        """
+        dir_exists = False
+        for mem_map_temp_dir in consts.UNIX_TEMP_DIRS:
+            # Check if the file already exists
+            file_path = os.path.join(mem_map_temp_dir,
+                                     consts.UNIX_TEMP_DIR_SUFFIX, mem_map_name)
+            if os.path.exists(file_path):
+                raise Exception(
+                    f'File {file_path} for memory map {mem_map_name} '
+                    f'already exists')
+            # Check if the parent directory exists
+            dir_path = os.path.join(mem_map_temp_dir,
+                                    consts.UNIX_TEMP_DIR_SUFFIX)
+            if os.path.isdir(dir_path):
+                dir_exists = True
+        # Check if any of the parent directories exists
+        if not dir_exists:
+            if not self._create_mem_map_dir():
+                return None
+        # Create the file
+        for mem_map_temp_dir in consts.UNIX_TEMP_DIRS:
+            file_path = os.path.join(mem_map_temp_dir,
+                                     consts.UNIX_TEMP_DIR_SUFFIX, mem_map_name)
+            try:
+                file = open(file_path, 'wb+')
+                file.truncate(mem_map_size)
+                return file
+            except Exception:
+                # If the memory map could not be created in this directory, we
+                # keep trying in other applicable directories.
+                pass
```

Comment:
These are the kind of things that may stay failing over and over without us knowing.
Should we make this a little smarter?
Is there an option to have counters in the worker? Is there a way for these counters to float to the C# side?

Reply:
Well, we don't care particularly about one specific directory not being created successfully. But if neither of them was created, then we will end up logging the error at the end of this function. So if we start seeing those in the logs, that means something is not right.